packages/slate-react/src/components/android/android-editable.tsx

```diff
@@ -0,0 +1,636 @@
+import React, {
+  useCallback,
+  useEffect,
+  useLayoutEffect,
+  useMemo,
+  useRef,
+  useState,
+} from 'react'
+import {
+  Descendant,
+  Editor,
+  Element,
+  Node,
+  NodeEntry,
+  Range,
+  Text,
+  Transforms,
+} from 'slate'
+import throttle from 'lodash/throttle'
+import scrollIntoView from 'scroll-into-view-if-needed'
+
+import {
+  IS_CHROME_LEGACY,
+  IS_EDGE_LEGACY,
+  IS_FIREFOX,
+  IS_SAFARI,
+} from '../../utils/environment'
+import { ReactEditor } from '../..'
+import { ReadOnlyContext } from '../../hooks/use-read-only'
+import { useSlate } from '../../hooks/use-slate'
+import { useIsomorphicLayoutEffect } from '../../hooks/use-isomorphic-layout-effect'
+import {
+  DOMElement,
+  DOMNode,
+  DOMRange,
+  DOMStaticRange,
+  getDefaultView,
+  isDOMNode,
+  isPlainTextOnlyPaste,
+} from '../../utils/dom'
+import {
+  EDITOR_TO_ELEMENT,
+  EDITOR_TO_RESTORE_DOM,
+  EDITOR_TO_WINDOW,
+  ELEMENT_TO_NODE,
+  IS_FOCUSED,
+  IS_READ_ONLY,
+  NODE_TO_ELEMENT,
+  PLACEHOLDER_SYMBOL,
+} from '../../utils/weak-maps'
+import { AndroidInputManager } from './android-input-manager'
+import { EditableProps } from '../editable'
+import { ErrorBoundary } from './ErrorBoundary'
+import useChildren from '../../hooks/use-children'
+
+// COMPAT: Firefox/Edge Legacy don't support the `beforeinput` event
+// Chrome Legacy doesn't support `beforeinput` correctly
+const HAS_BEFORE_INPUT_SUPPORT = !(
+  IS_FIREFOX ||
+  IS_EDGE_LEGACY ||
+  IS_CHROME_LEGACY
+)
+
+export const AndroidEditableNoError = (props: EditableProps): JSX.Element => {
+  return (
+    <ErrorBoundary>
+      <AndroidEditable {...props} />
+    </ErrorBoundary>
+  )
+}
+
+/**
+ * Editable.
+ */
+
+export const AndroidEditable = (props: EditableProps): JSX.Element => {
+  const {
+    autoFocus,
+    decorate = defaultDecorate,
+    onDOMBeforeInput: propsOnDOMBeforeInput,
+    placeholder,
+    readOnly = false,
+    renderElement,
+    renderLeaf,
+    style = {},
+    as: Component = 'div',
+    ...attributes
+  } = props
+  const editor = useSlate()
+  const ref = useRef<HTMLDivElement>(null)
+  const inputManager = useMemo(() => new AndroidInputManager(editor), [editor])
```

Comment:
Absolutely! As all the reconciliation happens outside of React's lifecycle events, it gets no benefit from hooks. Using hooks would adds a layer of complexity while not improving the performance nor maintainability of the work required.

I could add a new hook that merely manages the android input manager, but that extra structure does not make the code any clearer as the `android-editable` class is all about managing these dependendies.

Reply:
> @wleroux can you please un-resolve the discussions with the exception of the ones where you have removed/moved code? I don't think it's fair to viewers of this PR to mark them as resolved.

@clauderic Sorry about that, it was my assumption that anyone can unresolve an issue / posting new comments would unresolve that conversation.

Reply:
Hi @clauderic, I've opened up a separate issue to discuss a hook based Android implementation. I'm going to "resolve" these conversations, because we aren't going to be implementing your proposal in this PR, but it is not meant to mean that the issue of which approach is better is resolved. I've re-posted your summary in the issue here #4223 as well as adding some comments about why we won't be implementing that idea in this particular PR.